numpy/core/src/umath/ufunc_object.c

```diff
@@ -184,6 +184,45 @@ _find_array_method(PyObject *args, int nin, PyObject *method_name)
     return method;
 }
 
+/*
+ * Returns an incref'ed pointer to the proper __array_prepare__/__array_wrap__
+ * method for a ufunc output argument, given the output argument `obj`, and the
+ * method chosen from the inputs `input_method`.
+ */
+static PyObject *
+_get_output_array_method(PyObject *obj, PyObject *method,
+                         PyObject *input_method) {
+    if (obj == Py_None) {
```

Comment:
There are now three times in which you have the incref and return of `input_method`. How about inverting the logic, and move the next two lines to the very end, as the fall-back opion, i.e.,
```
PyObject *ometh;
if obj != Py_None` {
    if (PyArrayCheckExact(obj)) {
        return Py_RETURN_NONE;
    }
    ometh = ...
    if (ometh) {
        if(PyCallable_Check(ometh)) {
            return ometh;
        }
        else {
            Py_DECREF(ometh);
        }
    }
    else {
        PyErr_Clear();
    }
}
Py_XINCREF(input_method);
return input_method;
```